core/src/main/java/it/pagopa/selfcare/mscore/core/strategy/factory/OnboardingInstitutionStrategyFactory.java

```diff
@@ -0,0 +1,257 @@
+                            : List.of(input.getInstitution().getDigitalAddress());
+
+                    File logoFile = contractService.getLogoFile();
+
+                    emailService.sendAutocompleteMail(destinationMail, new HashMap<>(), logoFile, EmailService.PAGOPA_LOGO_FILENAME);
+                }
+            } catch (Exception e) {
+                onboardingDao.rollbackSecondStep(input.getToUpdate(), input.getToDelete(), input.getInstitution().getId(),
+                        input.getOnboardingRollback().getToken(), input.getOnboardingRollback().getOnboarding(), input.getOnboardingRollback().getProductMap());
+            }
+        };
+    }
+
+    private Consumer<OnboardingInstitutionStrategyInput> verifyAndFillInstitutionAttributeStrategy() {
+        return input -> {
+            input.getRequest().setTokenType(TokenType.INSTITUTION);
+            Institution institution = institutionService.retrieveInstitutionByExternalId(input.getRequest().getInstitutionExternalId());
+            OnboardingInstitutionUtils.checkIfProductAlreadyOnboarded(institution, input.getRequest());
+            checkIncompleteOnboarding(institution, input.getRequest());
+            OnboardingInstitutionUtils.validateOverridingData(input.getRequest().getInstitutionUpdate(), institution);
+            List<GeographicTaxonomies> geographicTaxonomies = getGeographicTaxonomy(input.getRequest());
+            List<InstitutionGeographicTaxonomies> institutionGeographicTaxonomies = new ArrayList<>();
+            if (!geographicTaxonomies.isEmpty()) {
+                institutionGeographicTaxonomies = geographicTaxonomies.stream()
+                        .map(geo -> new InstitutionGeographicTaxonomies(geo.getGeotaxId(), geo.getDescription()))
+                        .collect(Collectors.toList());
+            }
+
+            input.setInstitution(institution);
+            input.setInstitutionGeographicTaxonomies(institutionGeographicTaxonomies);
+        };
+    }
+
+    private void checkIncompleteOnboarding(Institution institution, OnboardingRequest request) {
+        List<Token> tokens = new ArrayList<>();
+        if (institution.getOnboarding() != null) {
+            tokens = institution.getOnboarding().stream()
+                    .filter(o -> o.getProductId().equalsIgnoreCase(request.getProductId())
+                            && (o.getStatus() == RelationshipState.PENDING || o.getStatus() == RelationshipState.TOBEVALIDATED))
+                    .map(o -> onboardingDao.getTokenById(o.getTokenId()))
+                    .collect(Collectors.toList());
+        }
+        var now = OffsetDateTime.now();
+        boolean isIncomplete = false;
+        for (var token : tokens) {
+            if (TokenUtils.isTokenExpired(token, now)) {
+                onboardingDao.persistForUpdate(token, institution, RelationshipState.DELETED, null);
+            } else {
+                isIncomplete = true;
```

Comment:
I'd add a comment on what's happening here

Reply:
yep, I mean in the code to let other devs understand what's happening. ie.

// we set the state to DELETED for all the expired tokens and...